File: 卡码/代码随想录/1/06-二叉树/Java/binaryTree/src/main/java/com/kurumi/binarytree/_02_二叉树属性/LeetCode_0513.java
//513. 找树左下角的值
package com.kurumi.binarytree._02_二叉树属性;

import com.kurumi.binarytree.TreeNode;
import java.util.LinkedList;
import java.util.Queue;

public class LeetCode_0513 {

}
class Solution {
  int maxDepth=-1;
  int res;
  public void traversal(TreeNode root,int depth){
    if(root.left==null&&root.right==null){
      if(depth>maxDepth){
        maxDepth=depth;
        res=root.val;
      }
      return;
    }
    if(root.left!=null) traversal(root.left,depth+1);
    if(root.right!=null) traversal(root.right, depth+1);
    return;
  }
  public int findBottomLeftValue(TreeNode root) {
    traversal(root, 0);
    return res;
  }
//  public int findBottomLeftValue(TreeNode root) {
//    Queue<TreeNode> q=new LinkedList<>();
//    if(root!=null) q.offer(root);
//    int res=0;
//    while(!q.isEmpty()){
//      int size=q.size();
//      for(int i=0;i<size;i++){
//        TreeNode node = q.poll();
//        if(i==0) res=node.val;
//        if(node.left!=null) q.offer(node.left);
//        if(node.right!=null) q.offer(node.right);
//      }
//    }
//    return res;
  }
}

File: 卡码/代码随想录/1/06-二叉树/Python/02-二叉树属性/LeetCode_0513.py
#513. 找树左下角的值
from typing import *
from ..TreeNode import *

class Solution:
  def __init__(self):
    self.maxDepth:int=-1
    self.res:int=0
  def traversal(self,root:Optional[TreeNode],depth:int):
    if not root.left and not root.right:
      if depth>self.maxDepth:
        self.maxDepth=depth
        self.res=root.val
      return
    root.left and self.traversal(root.left,depth+1)
    root.right and self.traversal(root.right,depth+1)

  def findBottomLeftValue(self, root: Optional[TreeNode]) -> int:
    self.traversal(root,0)
    return self.res
#   def findBottomLeftValue(self, root: Optional[TreeNode]) -> int:
#     q:List[TreeNode]=[]
#     if root:
#       q.append(root)
#     res:int=0
#     while len(q)>0:
#       size:int=len(q)
#       for i in range(size):
#         node:TreeNode=q[0]
#         q=q[1:]
#         if i==0:
#           res=node.val
#         if node.left:
#           q.append(node.left)
#         if node.right:
#           q.append(node.right)
#     return res
